[fix] [`redundant_closure_for_method_calls`] Suggest relative paths for local modules

modelflat opened the pull request:
Fixes #10854.

Currently, `redundant_closure_for_method_calls` suggest incorrect paths when a method defined on a struct within inline mod is referenced (see the description in the aforementioned issue for an example; also see [this playground link](https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=f7d3c5b2663c9bd3ab7abdb0bd38ee43) for the current-version output for the test cases added in this PR). It will now try to construct a relative path path to the module and suggest it instead.

changelog: [`redundant_closure_for_method_calls`] Fix incorrect path suggestions for types within local modules

modelflat commented:
@Centri3 Thanks for the suggestion! I moved `get_ucfs_type_name` to `lib.rs` in the utils. Considering the size of `lib.rs` I'd probably prefer to put it into a separate module, but I'm not sure which one is the most suitable... so I guess `lib.rs` it is 🙂

Centri3 commented on clippy_utils/src/lib.rs:
```diff
@@ -2884,3 +2885,111 @@ op_utils! {
+            // mod_a::sym  ::     ::
+            // (looking at mod_c)
+            EitherOrBoth::Right(r) => {
+                if let DefPathData::TypeNs(_) = r.data {
```
This would probably be better as a guard instead (same with the above)

Centri3 commented on clippy_utils/src/lib.rs:
```diff
@@ -2884,3 +2885,111 @@ op_utils! {
+                }
+                // Otherwise ignore
+            },
+            EitherOrBoth::Both(l, r) => match (l.data, r.data) {
```
afaict, this can be simplified by pushing `l` and `r` if they're `TypeNs`.

Centri3 commented on clippy_utils/src/lib.rs:
```diff
@@ -2884,3 +2885,111 @@ op_utils! {
+///
+/// TODO make use of `use` items. Maybe we should have something more sophisticated like
+/// rust-analyzer does? <https://docs.rs/ra_ap_hir_def/0.0.169/src/ra_ap_hir_def/find_path.rs.html#19-27>
+fn get_relative_path(from: &DefPath, to: &DefPath) -> String {
```
This should document that `from` and `to` have to be local.

Centri3 commented on clippy_utils/src/lib.rs:
```diff
@@ -2884,3 +2885,111 @@ op_utils! {
+            match ty.kind() {
+                rustc_ty::Adt(adt, _) => get_path_to_callee(tcx, from, adt.did()),
+                rustc_ty::Array(..)
+                | rustc_ty::Dynamic(..)
+                | rustc_ty::Never
+                | rustc_ty::RawPtr(_)
+                | rustc_ty::Ref(..)
+                | rustc_ty::Slice(_)
+                | rustc_ty::Tuple(_) => format!("<{}>", EarlyBinder::bind(ty).instantiate(tcx, args)),
```
We need to handle these too, I'm pretty sure. This should recursively go through each nested type and get the relative path for each.

Not an issue for the lint in question, but if somebody uses this on something like `(super::A, super::super::B)` it'll still be broken afaik.

bors commented:
:umbrella: The latest upstream changes (presumably #11398) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

blyxyas commented:
Seeing as Catherine is pretty busy (judging from her GitHub contribution graph) so I'll take from now on some of her reviews.

r? @blyxyas

blyxyas commented on clippy_utils/src/lib.rs:
```diff
@@ -2884,3 +2886,116 @@ op_utils! {
+            Right(DefPathData::TypeNs(_)) => go_up_by += 1,
+            _ => {},
+        }
+    }
+
+    repeat(String::from("super")).take(go_up_by).chain(path).join("::")
```
Seeing the result, I think that we should maybe cap `go_up_by` at like 2, AFAIK people prefer having `crate::a::b` than `super::super::super::super::b` in their codebase.

modelflat replied:
I really like this idea, implemented!

blyxyas reviewed (commented):
Very good first iteration.

blyxyas commented:
@modelflat Are you there?

modelflat commented:
@blyxyas yep, hopefully I'll get some time to work on it in December

blyxyas commented:
@modelflat Would you like for other person to finish this contribution?

modelflat commented:
@blyxyas Hi! Sorry I completely forgot about this one. I've got some spare time now so can work on it. Pls take a look at the newest changes when you have time, I resolved conflicts and addressed your comments

blyxyas reviewed (approved):
LGTM! Could you squash these commits? ฅ*•ω•*ฅ♡

blyxyas commented on tests/ui/eta.rs:
```diff
@@ -417,3 +417,57 @@ fn _closure_with_types() {
     let _ = f2(|x| -> u32 { f(x) });
 }
+
+/// https://github.com/rust-lang/rust-clippy/issues/10854.
```
(mini-fix)

```suggestion
/// https://github.com/rust-lang/rust-clippy/issues/10854
```

blyxyas commented:
@bors r+

bors commented:
:pushpin: Commit b3d53774e7ffcde40b18e18be700f63204ec96bd has been approved by `blyxyas`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=blyxyas b3d53774e7ffcde40b18e18be700f63204ec96bd -->
<!-- homu: {"type":"Approved","sha":"b3d53774e7ffcde40b18e18be700f63204ec96bd","approver":"blyxyas","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit b3d53774e7ffcde40b18e18be700f63204ec96bd with merge 85d808273c317e7c6a905c2bf24b8a9e9609bdc0...
<!-- homu: {"type":"BuildStarted","head_sha":"b3d53774e7ffcde40b18e18be700f63204ec96bd","merge_sha":"85d808273c317e7c6a905c2bf24b8a9e9609bdc0"} -->

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/7695477878/job/20968634394)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/actions/runs/7695477878/job/20968634394","builder_name":"checks-action_test"} -->

blyxyas commented:
@bors retry

bors commented:
:hourglass: Testing commit b3d53774e7ffcde40b18e18be700f63204ec96bd with merge 813d6728d8a20923ffbbc55d5090e4ab3f7f351f...
<!-- homu: {"type":"BuildStarted","head_sha":"b3d53774e7ffcde40b18e18be700f63204ec96bd","merge_sha":"813d6728d8a20923ffbbc55d5090e4ab3f7f351f"} -->

bors commented:
:boom: Test timed out
<!-- homu: {"type":"TimedOut"} -->

blyxyas commented:
@bors retry

bors commented:
:hourglass: Testing commit b3d53774e7ffcde40b18e18be700f63204ec96bd with merge 3cd713a9f8aee9651f090806126d633967ffbf68...
<!-- homu: {"type":"BuildStarted","head_sha":"b3d53774e7ffcde40b18e18be700f63204ec96bd","merge_sha":"3cd713a9f8aee9651f090806126d633967ffbf68"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/7697965079/job/20976215600), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/7697965080/job/20976222003), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/7697965065/job/20976716545)
Approved by: blyxyas
Pushing 3cd713a9f8aee9651f090806126d633967ffbf68 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"blyxyas","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7697965065/job/20976716545","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7697965079/job/20976215600","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7697965080/job/20976222003"},"merge_sha":"3cd713a9f8aee9651f090806126d633967ffbf68"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/7697965079/job/20976215600), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/7697965080/job/20976222003), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/7697965065/job/20976716545)
Approved by: blyxyas
Pushing 3cd713a9f8aee9651f090806126d633967ffbf68 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"blyxyas","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7697965065/job/20976716545","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7697965079/job/20976215600","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7697965080/job/20976222003"},"merge_sha":"3cd713a9f8aee9651f090806126d633967ffbf68"} -->

bors commented:
:eyes: Test was successful, but fast-forwarding failed: 422 Changes must be made through a pull request.
